# Assignment  week 2

## Notebook made by   (If not filled in correctly: 0 pts for assignment)

__Name(s)__: Adriaan de Vries, Verna Dankers

__Student id(s)__ : 10795227, 10761225

__Email(s)__: adriaan.devries@student.uva.nl, verna.dankers@student.uva.nl

### Pledge (taken from [Coursera's Honor Code](https://www.coursera.org/about/terms/honorcode) )



Put here a selfie with your photo where you hold a signed paper with the following text: (if this is team work, put two selfies here). The link must be to some place on the web, not to a local file. **Assignments without the selfies will not be graded and receive 0 points.**

> My answers to homework, quizzes and exams will be my own work (except for assignments that explicitly permit collaboration).

>I will not make solutions to homework, quizzes or exams available to anyone else. This includes both solutions written by me, as well as any official solutions provided by the course staff.

>I will not engage in any other activities that will dishonestly improve my results or dishonestly improve/hurt the results of others.

<img width=30% src='http://oi63.tinypic.com/2q3srh3.jpg'/>
<img src="http://i63.tinypic.com/2itj1ic.jpg" border="0">

## Questions

#### Book

* Chap 6: 6.8 to 6.13, 6.14 to 6.17. 
 
#### Programming

* [MakeInvertedIndex](MakeInvertedIndex1516.html): Exercises 8.1 to 8.6 about Boolean Search Engines and Exercises 9.1 to 9.5 about Ranked Retrieval 
 


# Book

#### 6.8 

$idf_t = log( \frac{N}{df_t} )$

For $df_t > 0$, the $idf_t$ will always stay between the lower and the upper bound. The lower bound for $idf_t$ will be reached when the $df_t = N$ and the upper bound will be reached when $df_t = 1$. So the $idf_t$ can only be one of the numbers from the finite set of the number between these two bounds.

#### 6.9 

The $idf_t$ of a term that appears in every document: $log(\frac{N}{N}) = log(1) = 0$

Such a word will act like a stop word when the $tf-idf$ weighting will be used, since its weight becomes zero. It will not be able to affect the search results, due to its zero weight.

#### 6.10

$tf idf_{t,d} = tf_{t,d} \times idf_t$

##### car
$tf idf_{car,d1} = 27 \times log(\frac{3}{3}) = 0 $  

$tf idf_{car,d2} = 4 \times log(\frac{3}{3}) = 0 $  

$tf idf_{car,d3} = 24 \times log(\frac{3}{3}) = 0 $  

##### auto

$tf idf_{auto,d1} = 3 \times log(\frac{3}{2}) = 0.53 $  

$tf idf_{auto,d2} = 33 \times log(\frac{3}{2}) = 5.81 $

$tf idf_{auto,d3} = 0 \times log(\frac{3}{2}) = 0 $  

##### insurance

$tf idf_{insurance,d1} = 0 \times log(\frac{3}{2}) = 0 $  

$tf idf_{insurance,d2} = 33 \times log(\frac{3}{2}) = 5.81 $  

$tf idf_{insurance,d3} = 29 \times log(\frac{3}{2}) = 5.11 $  

##### best

$tf idf_{best,d1} = 14 \times log(\frac{3}{2}) = 2.47 $  

$tf idf_{best,d2} = 0 \times log(\frac{3}{2}) = 0 $  

$tf idf_{best,d3} = 17 \times log(\frac{3}{2}) = 2.99 $  


#### 6.11

Yes, obivously. Dependend on the number of documents in the collection (N) the upper bound for $tf-idf$ is $log(\frac{N}{1})$, which can definitely exceed one. This is illustrated by exercise 6.10. 

#### 6.12

Altering the base of a logarithm is equivalent to multiplying the result by a constant. This will obviously alter the specific values for the idf, and thus also for $Score(q,d)$ (6.9), but the ratios between scores of different documents will remain the same. For sintance, if in base A the score of Doc1 is 2, and it's 4 in base B, than Doc2 will also have a score twice as high in base B as opposed to A. It is, however, important to calculate the score for each document with a logarithm of the same base. If we change bases during the process, some documents could get unfairly high or low rankings. There is, however, absolutely no reason to change the base of the logarithm mid-computation.

#### 6.13

As stated before, the result of a logarithm in base two for some value, is just a constant times the logarithm with another base of that value. There really cannot be an easy approximation that works in base 2 but not in other bases. However, if what is meant in the question is an easy approximation of the idf in base 2 given the idf in base 10, then one could get this value by multiplying the idf by $\frac{log(10)}{log(2)}$. For these logarithms, any base gives literally the same exact solution, as they're divided by eachother. Also, this is not so much an approximation as it is an exact way to calculate the idf in base 2 from base 10.

##### 6.15

Both the term frequency and the inverse document frequency would have to see jealous and jealousy as the same term. For the term frequency, it's sufficient to simply add the term frequencies of the two previously seperate terms. In the calculation of the inverse document frequency, we can't simply add the document frequencies of the two terms as there may be overlap. Instead, we'd have to take the sets of documents that the terms appear in, take the union of those sets, and the cardinality of this set is the new document frequency of jealous(y) that can be used normally for the calculation of the idf.

#### 6.15

Vectors from exercise 6.10:

1. d1 = $[0,0.53,0,2.47]^T$
2. d2 = $[0,5.81,5.81,0]^T$
3. d3 = $[0,0,5.11,2.99]^T$


Normalized vectors:

1. d1 = $\frac{ [0,0.53,0,2.47]^T }{\sqrt{0.53^2 + 2.47^2}} = \frac{[0,0.53,0,2.47]^T}{2.53} = [0,0.21, 0  , 0.98]^T$
2. d2 = $\frac{ [0,5.81,5.81,0]^T }{\sqrt{5.81^2 + 5.81^2}} = \frac{[0,5.81,5.81,0]^T}{8.22} = [0,0.71,0.71, 0   ]^T$
3. d3 = $\frac{ [0,0,5.11,2.99]^T }{\sqrt{5.11^2 + 2.99^2}} = \frac{[0,0,5.11,2.99]^T}{5.92} = [0,   0,0.86, 0.51]^T$

#### 6.16

1. length d1: $\sqrt{0.21^2 + 0.98^2} = 1.002$
2. length d2: $\sqrt{0.71^2 + 0.71^2} = 1.004$
3. length d3: $\sqrt{0.86^2 + 0.51^2} = 1.000$ 

Now the vectors have length 1 since we divided the original vectors by their length in exercise 6.15.

#### 6.17

1. The weight of a term is 1 if present in the query, 0 otherwise.

query: car insurance

1. d1 = $[0,0.21, 0  , 0.98]^T \cdot [1,0,1,0]^T = 0$
2. d2 = $[0,0.71,0.71, 0   ]^T \cdot [1,0,1,0]^T = 0.71$
3. d3 = $[0,   0,0.86, 0.51]^T \cdot [1,0,1,0]^T = 0.86$

2. Euclidean normalized idf.

1. d1 = $[0,0.21, 0  , 0.98]^T \cdot [0,0,0.18,0]^T = 0$
2. d2 = $[0,0.71,0.71, 0   ]^T \cdot [0,0,0.18,0]^T = 0.13$
3. d3 = $[0,   0,0.86, 0.51]^T \cdot [0,0,0.18,0]^T = 0.15$

If we normalize the query, the dot product will be $\frac{1}{0.18}$ times as high for all documents.

For both ways of calculating the third document is the best hit for the query, followed by the second document.

# Programming

For the programming part, we'll first copy and paste what we need from Assignment 1.

In [2]:
from bs4 import BeautifulSoup
from nltk.tokenize import WhitespaceTokenizer
from collections import Counter, defaultdict
import os
import pickle
import re

CORPUS = "Shakespeare"

def normalize(word, keep_capitals):
    """
    Normalize capital letters and punctuation
    """
    if (keep_capitals and not word == word.upper()) or not keep_capitals:
        word = word.lower()
    word = re.sub(r'[^\w\s]', '', word)
    return word

def prepare_data(folder, new_folder, keep_capitals, byte):
    """
    Prepare a text corpus for the process of
    making an inverted index.
    """
    # Create a directory for the processed files
    if not os.path.exists(new_folder):
        os.makedirs(new_folder)

    # Prepare each file one by one
    for filename in os.listdir(folder):
        if not os.path.exists(os.path.join(new_folder, filename[:-4] + ".pkl")):
            # Strip xml tags
            if byte:
                xmlfile = BeautifulSoup(open(os.path.join(folder, filename), 'rb'),"xml")
            else:
                xmlfile = BeautifulSoup(open(os.path.join(folder, filename), 'r'),"xml")
            textfile = xmlfile.get_text()

            # Split the text on whitespaces
            beautifile = WhitespaceTokenizer().tokenize(textfile)

            # Normalize text for capital letters, non-alphabetical tokens and whitespace
            finalfile = [ normalize(word, keep_capitals)
                          for word 
                          in beautifile 
                         ]
            new_filename = os.path.join(new_folder, filename[:-4] + ".pkl")
            savefile = open(new_filename, 'wb')
            pickle.dump(finalfile, savefile)


In [3]:
def index_collection(folder):
    """
    Create an inverted index given a folder 
    containig a text corpus.
    """
    index= defaultdict(Counter) # initialize MyIndex
    for filename in os.listdir(folder):
        text = pickle.load(open(os.path.join(folder,filename),'rb'))
        
        # Update the index with each token
        for w in text:    
            index[w][filename[:-4]]+=1   
    if '' in index.keys():
        del index['']
    return index

In [4]:
def calculate_doc_freq(index, word):
    """
    Calculate in how many documents a given 
    term occurs, based on the inverted index.
    """
    return len(index[word])
    
def calculate_corp_freq(index, word):
    """
    Calculate how often a term occurs in a 
    corpus, based on the inverted index.
    """
    corpus_frequency = 0
    
    for document in index[word]:
        corpus_frequency += index[word][document]
        
    return corpus_frequency

def add_frequencies(index):
    for word in index:
        original_counter = index[word]
        document_frequency = calculate_doc_freq(index, word)
        corpus_frequency = calculate_corp_freq(index, word)
        index[word] = {"counter" : original_counter, 
                       "doc_freq" : document_frequency, 
                       "corp_freq" : corpus_frequency
                      }
    return index

In [5]:
prepare_data(CORPUS, CORPUS + "New", 0, 0)
index_org = index_collection(CORPUS + "New")
index_with_freq = add_frequencies(index_org)

#### Making a list of all documents

Will be useful for negation.

In [6]:
def make_doc_set(index):
    """
    Create a set containing all documents that
    are present in a given index.
    """
    doc_list = []
    for token in index:
        doc_list.extend(list(index[token]['counter'].elements()))
    return set(doc_list)

In [7]:
doc_set = make_doc_set(index_with_freq)
print(doc_set)

{'rich_iii', 'all_well', 'john', 'two_gent', 'lll', 'macbeth', 'timon', 'pericles', 'hen_iv_1', 'much_ado', 'cymbelin', 'hen_iv_2', 'rich_ii', 'win_tale', 'taming', 'dream', 't_night', 'coriolan', 'r_and_j', 'merchant', 'tempest', 'troilus', 'hen_vi_3', 'hen_vi_2', 'hen_vi_1', 'com_err', 'othello', 'hen_viii', 'j_caesar', 'hen_v', 'titus', 'hamlet', 'm_wives', 'm_for_m', 'lear', 'as_you', 'a_and_c'}


### On to the actual exercises:

We start with the code and the results for the Shakespeare corpus.

#### 8.1 Create a search engine for negation (very useful)

In [8]:
def docs_with_token(index, token):
    """
    Return all documents that contain a
    certain token.
    """
    if token in index:
        return set(index[token]['counter'].elements())
    else:
        return set()

In [9]:
def negation_search(doc_set, index, word):
    """
    Find all documents form an index that
    do not contain a certain word.
    """
    return doc_set.difference(docs_with_token(index, word))

In [10]:
print (negation_search(doc_set, index_with_freq, 'test'))
print (docs_with_token(index_with_freq, 'test'))
print (negation_search(doc_set, index_with_freq, 'despair'))

{'rich_iii', 'merchant', 'john', 'two_gent', 'lll', 'timon', 'pericles', 'hen_iv_1', 'much_ado', 'cymbelin', 'hen_iv_2', 'rich_ii', 'win_tale', 'taming', 'dream', 't_night', 'coriolan', 'r_and_j', 'all_well', 'm_wives', 'hen_vi_3', 'hen_vi_2', 'hen_vi_1', 'com_err', 'hen_viii', 'j_caesar', 'hen_v', 'titus', 'macbeth', 'troilus', 'lear', 'as_you', 'a_and_c'}
{'m_for_m', 'tempest', 'othello', 'hamlet'}
{'othello', 'm_wives', 'hen_v', 't_night', 'hamlet', 'dream', 'two_gent', 'lll', 'j_caesar', 'troilus', 'timon', 'hen_viii', 'pericles', 'as_you', 'hen_iv_1', 'much_ado'}


#### 8.2 Disjunction search

Input as a list of words. Output is unordered.

In [11]:
def disjunction_search(query, index):
    """
    Return documents that contain at least
    one of the words from the query.
    """
    if len(query) ==0:
        return set()
    result = docs_with_token(index, query[0])
    for word in query[1:]:
        result = result.union(docs_with_token(index, word))
    return result

In [12]:
print(disjunction_search(['test', 'despair'], index_with_freq))

{'rich_iii', 'tempest', 'merchant', 'hen_vi_3', 'hen_vi_2', 'hen_vi_1', 'com_err', 'john', 'rich_ii', 'macbeth', 'cymbelin', 'hen_iv_2', 'othello', 'm_for_m', 'win_tale', 'taming', 'hamlet', 'all_well', 'titus', 'coriolan', 'lear', 'r_and_j', 'a_and_c'}


##### 8.3.1 Ordering by number of terms matched

In [13]:
def num_of_terms(query, index, doc):
    """
    Return how many words from the query
    a given document contains.
    """
    count = 0
    for token in query:
        if doc in docs_with_token(index, token):
            count += 1
    return count

def disj_search_simple_order(query, index):
    """
    Return documents that contain words from the query
    ordered by the number of query words they contained.
    """
    set_of_docs = disjunction_search(query, index)
    return sorted(set_of_docs, reverse = True, key = lambda x: num_of_terms(query,index,x))

In [14]:
print(disj_search_simple_order(['test', 'despair', 'fat', 'beast'], index_with_freq))
print('merchant:', num_of_terms(['test', 'despair', 'fat', 'beast'], index_with_freq, 'merchant'))
print('r_and_j:', num_of_terms(['test', 'despair', 'fat', 'beast'], index_with_freq, 'r_and_j'))
print('all_well:', num_of_terms(['test', 'despair', 'fat', 'beast'], index_with_freq, 'all_well'))

['rich_iii', 'merchant', 'john', 'hamlet', 'hen_iv_2', 'taming', 'tempest', 'com_err', 'm_for_m', 'titus', 'a_and_c', 'two_gent', 'lll', 'hen_iv_1', 'othello', 'win_tale', 'dream', 'coriolan', 'r_and_j', 'troilus', 'hen_vi_3', 'hen_vi_2', 'hen_vi_1', 'j_caesar', 'hen_v', 'macbeth', 'm_wives', 'lear', 'as_you', 'timon', 'much_ado', 'cymbelin', 't_night', 'all_well', 'rich_ii']
merchant: 3
r_and_j: 2
all_well: 1


##### 8.3.2 Ordering by term frequency

In [15]:
def term_frequency(query, index, doc):
    """
    Count how often the terms from a query
    appear in a given document.
    """
    count = 0
    for token in query:
        if doc in index[token]['counter']:
            count += index[token]['counter'][doc]
    return count

def disj_search_term_freq_order(query, index):
    """
    Return documents that contain words from the query
    ordered by how frequent those terms appeared.
    """
    set_of_docs = disjunction_search(query, index)
    return sorted(set_of_docs, reverse = True, key = lambda x: term_frequency(query,index,x))

In [16]:
for doc in disj_search_term_freq_order(['test', 'despair', 'fat', 'beast'], index_with_freq):
    print (doc, term_frequency(['test', 'despair', 'fat', 'beast'], index_with_freq, doc))

hen_iv_1 19
rich_iii 17
m_wives 16
hamlet 13
timon 11
lll 8
hen_iv_2 8
taming 7
com_err 7
rich_ii 6
othello 5
r_and_j 5
hen_vi_3 5
lear 5
as_you 5
john 4
dream 4
tempest 4
troilus 4
m_for_m 4
merchant 3
win_tale 3
hen_vi_2 3
j_caesar 3
hen_v 3
macbeth 3
titus 3
a_and_c 3
two_gent 2
much_ado 2
cymbelin 2
coriolan 2
all_well 2
hen_vi_1 2
t_night 1


##### 8.3.3 Ordering by number of matched terms and term frequency

In [17]:
def tf_and_matchedNo_combo(query, index, doc):
    """
    Returns both the sum of term frequencies for the query and document, 
    as the number of terms from the query that match the document
    """
    return([num_of_terms(query, index, doc), term_frequency(query, index, doc)])

def disj_search_combo_order(query, index):
    """
    Return documents that contain words from the query
    ordered by how many words from the query they
    contained and on how frequent those terms appeared.
    """
    set_of_docs = disjunction_search(query, index)
    return sorted(set_of_docs, 
                  reverse = True, 
                  key = lambda x: tf_and_matchedNo_combo(query, index, x)
                 )

In [18]:
for doc in disj_search_combo_order(['test', 'despair', 'fat', 'beast'], 
                                   index_with_freq):
    print (doc, tf_and_matchedNo_combo(['test', 'despair', 'fat', 'beast'], 
                                       index_with_freq, doc))

rich_iii [3, 17]
hamlet [3, 13]
hen_iv_2 [3, 8]
taming [3, 7]
com_err [3, 7]
john [3, 4]
tempest [3, 4]
m_for_m [3, 4]
merchant [3, 3]
titus [3, 3]
a_and_c [3, 3]
hen_iv_1 [2, 19]
m_wives [2, 16]
lll [2, 8]
othello [2, 5]
r_and_j [2, 5]
hen_vi_3 [2, 5]
lear [2, 5]
as_you [2, 5]
dream [2, 4]
troilus [2, 4]
win_tale [2, 3]
hen_vi_2 [2, 3]
j_caesar [2, 3]
hen_v [2, 3]
macbeth [2, 3]
two_gent [2, 2]
coriolan [2, 2]
hen_vi_1 [2, 2]
timon [1, 11]
rich_ii [1, 6]
much_ado [1, 2]
cymbelin [1, 2]
all_well [1, 2]
t_night [1, 1]


##### 8.4 Conjunctive search

Input is, again, a list of words.

In [19]:
def conjunction_search(query, index):
    """
    Return all documents that contain all words
    from the given query.
    """
    result = ""
    
    if len(query)==0:
        return set()
    
    for word in query:
        if result:
            result = result.intersection(docs_with_token(index, word))
        else: 
            result = docs_with_token(index, word)
    return result

In [20]:
print(conjunction_search(['despair','terrible'], index_with_freq))

{'rich_iii', 'tempest', 'rich_ii', 'all_well', 'hen_vi_2', 'hen_vi_1', 'macbeth', 'coriolan', 'lear', 'a_and_c', 'cymbelin', 'hen_iv_2'}


##### 8.5 Ordering of conjunctive search

Conjunctive search could be ordered by the amount of times terms appear in a document. Simply taking the sum is an option, or one could choose to do it relative to the rarity of the tokens. So the ranking could be the total amount of times a word from the query appears in the document, divided by the total amount of times the word appears in the corpus. Taking the sum over all the words in the query in this way would give a fairly good ranking.

Each document score could then possibibly be divided by the document length, to punish longer documents as they will obviously on average have more terms in them. One could, however, argue that these longer documents are better documents, so they should be ranked higher and not punished.

##### 8.6 Relaxing the search

Instead of removing documents that fail to contain a search word, we can add some form of counter to them. Only when the counter exceeds a specified value will it be deleted. Then what's left is to decide how many terms the document should at least match. We think three quarters would be a good threshold, so it would not be allowed to miss more than floor(N/4) where N is the amount of words in the query. This is easily adjustable to any other value.

As for ranking, the documents could be ranked the same way as specified in 8.5, except the documents with the lowest counter would be placed highest. So the documents that match the most terms will end up at the top of the ranking.

#### 9.1 TF-IDF

In [21]:
from math import log

def TF_IDF(term, doc_set, index):
    """
    Return the tf-idf weight of a given term.
    """
    rated_docs = []
    IDF = log(len(doc_set) / len(docs_with_token(index, term)))
    for doc in doc_set:
        if doc in index[term]['counter']:
            rated_docs.append([index[term]['counter'][doc]*IDF,doc])
        else:
            rated_docs.append([0,doc])
    return sorted(rated_docs, reverse=True)

In [22]:
TF_IDF('despair', doc_set, index_with_freq)

[[7.9295366488912205, 'rich_iii'],
 [3.398372849524809, 'rich_ii'],
 [2.265581899683206, 'hen_vi_3'],
 [1.6991864247624044, 'lear'],
 [1.132790949841603, 'win_tale'],
 [1.132790949841603, 'r_and_j'],
 [1.132790949841603, 'macbeth'],
 [1.132790949841603, 'john'],
 [1.132790949841603, 'hen_vi_2'],
 [1.132790949841603, 'cymbelin'],
 [1.132790949841603, 'all_well'],
 [0.5663954749208014, 'titus'],
 [0.5663954749208014, 'tempest'],
 [0.5663954749208014, 'taming'],
 [0.5663954749208014, 'merchant'],
 [0.5663954749208014, 'm_for_m'],
 [0.5663954749208014, 'hen_vi_1'],
 [0.5663954749208014, 'hen_iv_2'],
 [0.5663954749208014, 'coriolan'],
 [0.5663954749208014, 'com_err'],
 [0.5663954749208014, 'a_and_c'],
 [0, 'two_gent'],
 [0, 'troilus'],
 [0, 'timon'],
 [0, 't_night'],
 [0, 'pericles'],
 [0, 'othello'],
 [0, 'much_ado'],
 [0, 'm_wives'],
 [0, 'lll'],
 [0, 'j_caesar'],
 [0, 'hen_viii'],
 [0, 'hen_v'],
 [0, 'hen_iv_1'],
 [0, 'hamlet'],
 [0, 'dream'],
 [0, 'as_you']]

#### 9.2 TF-IDF scoring

In [23]:
def getKey(list_):
    return list_[1]

def TF_IDF_search(query, doc_set, index):
    """
    For all documents calculate how they
    score compared to the query, by tf-idf weighting.
    """
    dic = {}
    for doc in doc_set:
        dic[doc]=0
    for term in query:
        TFIDF = TF_IDF(term, doc_set, index)
        for elem in TFIDF:
            if elem[0]==0:
                break
            else:
                dic[elem[1]]+=elem[0]
    return sorted(dic.items(), reverse=True, key=getKey)

In [24]:
returnval = TF_IDF_search(['test', 'despair', 'fat', 'beast'], doc_set, index_with_freq)
for elem in returnval:
    print (elem)

('hen_iv_1', 9.793193846706075)
('rich_iii', 8.983176289127522)
('m_wives', 7.448460637417568)
('hamlet', 7.084728896995646)
('lll', 4.20839040710336)
('hen_iv_2', 4.20839040710336)
('rich_ii', 3.398372849524809)
('tempest', 3.278263191760636)
('m_for_m', 3.278263191760636)
('othello', 3.199111882155336)
('com_err', 2.9964481476564573)
('taming', 2.9964481476564573)
('timon', 2.679842909235256)
('hen_vi_3', 2.5092039823409564)
('lear', 2.1864305900779053)
('john', 1.9428085074201549)
('troilus', 1.9428085074201549)
('as_you', 1.8636571978148544)
('r_and_j', 1.8636571978148544)
('win_tale', 1.6991864247624044)
('hen_vi_2', 1.3764130324993533)
('merchant', 1.3764130324993533)
('titus', 1.3764130324993533)
('macbeth', 1.3764130324993533)
('a_and_c', 1.3764130324993533)
('dream', 1.2972617228940528)
('all_well', 1.132790949841603)
('hen_vi_1', 1.132790949841603)
('cymbelin', 1.132790949841603)
('hen_v', 1.0536396402363024)
('j_caesar', 1.0536396402363024)
('two_gent', 0.810017557578552)
('

#### 9.3 Euclidean length of docs

Function will return a dict, as that seems useful.

In [25]:
from math import sqrt

def euclidean_lengths(doc_set, index):
    """
    Compute for each document its Euclidean length.
    """
    dic={}
    for doc in doc_set:
        dic[str(doc)] = 0
    for key in index:
        for doc in index[key]['counter']:
            dic[str(doc)] += index[key]['counter'][str(doc)]**2
    for key in dic:
        dic[key] = sqrt(dic[key])
    return dic

In [26]:
print(euclidean_lengths(doc_set, index_with_freq))

{'rich_iii': 2529.0595089874814, 'tempest': 1355.7916506602332, 'm_wives': 2020.1821700034875, 'all_well': 2011.9169962997976, 'hen_vi_3': 2102.5584415183325, 'hen_vi_2': 2173.9512414035416, 'hen_vi_1': 1822.1355602698718, 'com_err': 1462.9962405966735, 'merchant': 1917.1390142605726, 'two_gent': 1501.6590824817729, 'othello': 2297.0711786969077, 'lll': 1852.4702426759789, 'john': 1771.741798344217, 'timon': 1492.1628597442036, 'pericles': 1527.190230455918, 'hen_iv_1': 2173.7451552562457, 'much_ado': 1900.3881182537425, 'cymbelin': 2259.077909236421, 'hen_iv_2': 2310.8329667026997, 'rich_ii': 1956.5198695643242, 'hamlet': 2665.07917330799, 'hen_v': 2310.7985632676855, 'titus': 1752.0644965297367, 'as_you': 1955.6847905529153, 'm_for_m': 1917.6334373388465, 'win_tale': 2088.30337834329, 'taming': 1843.2970460563322, 'macbeth': 1460.7621298486622, 'dream': 1387.939840194812, 'lear': 2162.9942209816463, 'troilus': 2117.9992917845843, 't_night': 1792.8792485831275, 'coriolan': 2372.472128

#### 9.4 Cosine Similarity

*If a word appears multiple times in the query, the length of the query vector is not just the square root of the length of the query. However, as this would change all results by multiplying them with a constant, it would not interfere with the ranking.

In [27]:
from math import sqrt

def cosine_similarity_search(query, doc_set, index):
    """
    Calculate the cosine similarity for each document
    in the set to the given query.
    """
    lengths = euclidean_lengths(doc_set, index)
    
    # for querylength we assume the query to consist of unique words, 
    # for justification check markdown box above
    querylength = sqrt(len(query))
    
    TF_IDF_results = TF_IDF_search(query, doc_set, index)
    final_result = []
    for result in TF_IDF_results:
        final_result.append([result[0],result[1] / (lengths[result[0]] * querylength)])
    
    # sort again
    return sorted(final_result, reverse=True, key=getKey)

In [28]:
result = cosine_similarity_search(['test', 'despair', 'fat', 'beast'], 
                                  doc_set, index_with_freq)
for [doc,similarity] in result:
    print (doc, similarity)

hen_iv_1 0.00225260855050684
m_wives 0.001843512121831248
rich_iii 0.001775991481656351
hamlet 0.0013291779411194436
tempest 0.0012089848724780876
lll 0.001135886102284738
com_err 0.0010240792370164859
hen_iv_2 0.0009105786674638504
timon 0.000897972661541332
rich_ii 0.0008684738914206772
m_for_m 0.0008547679467641046
taming 0.0008127957873277263
othello 0.0006963458319933607
hen_vi_3 0.0005967025536110594
john 0.0005482764218905397
lear 0.0005054175755230688
as_you 0.0004764717726541089
macbeth 0.00047112839399866986
r_and_j 0.000469102065731882
dream 0.00046733355629879767
troilus 0.00045864238835112707
win_tale 0.00040683418951091696
titus 0.00039279747841063344
merchant 0.0003589758025529063
a_and_c 0.0003336100656749558
hen_vi_2 0.00031656943501886376
hen_vi_1 0.00031084156814156793
j_caesar 0.0003072550832003015
all_well 0.00028152029927799383
two_gent 0.0002697075411550291
cymbelin 0.00025071976163595254
hen_v 0.00022798171527905862
coriolan 0.00017071171203333278
t_night 0.0001

#### 9.5 Okapi BM25

We will need to do some calculations with the values of k1, b, the term frequency, the document vector length and the average document vector length. All of these are calculable except for k1 and b, which are free variables. The wikipedia page suggests putting b=0.75 and k1 being between 1.2 and 2, so we'll take b=0.75 and k1=1.6.

In [29]:
from math import sqrt

def Okapi_BM25_search(query, doc_set, index):
    """
    Performs a search using the given query, using the 
    Okapi BM25 algorithm to rank the results
    """
    k1 = 1.6
    b = 0.75
    
    lengths = euclidean_lengths(doc_set, index)
    querylength = sqrt(len(query))
    
    # calculating the average doc length
    s = 0
    counter = 0
    for key in lengths:
        s += lengths[key]
        counter += 1
    avg_len = s/counter
    
    result = []
    for doc in doc_set:
        subresult = [str(doc), 0]
        for term in query:
            if doc in index[term]['counter']:
                tf = index[term]['counter'][str(doc)]
                val = (tf * (k1+1)) / (tf + k1 * (1-b+b*(lengths[str(doc)]/avg_len)))
                subresult[1]+=val
        result.append(subresult)
    return sorted(result, reverse=True, key=getKey)

In [30]:
result = Okapi_BM25_search(['test', 'despair', 'fat', 'beast'], 
                           doc_set, index_with_freq)
for [doc, Okapi_score] in result:
    print (doc, Okapi_score)

com_err 4.757900788908586
hamlet 4.735788265712867
rich_iii 4.475000185681661
taming 4.466287992275214
m_wives 4.218962695507707
hen_iv_1 3.991643004338016
tempest 3.935030785476477
hen_iv_2 3.8369575226705255
john 3.5779035786569833
m_for_m 3.467881609484815
lll 3.154760945652675
titus 3.1474160300055045
as_you 3.1373801507638284
r_and_j 3.12291745368682
lear 3.04234871003717
merchant 3.0234778718563793
dream 2.9869408622156985
a_and_c 2.9218776135644067
hen_vi_3 2.7916473899211267
othello 2.712978181281462
macbeth 2.707204937926539
troilus 2.6201123101368435
j_caesar 2.5639824550295582
win_tale 2.3793607597775392
hen_vi_2 2.340822056034838
timon 2.321752809048462
hen_v 2.281829347310559
two_gent 2.237402638211719
hen_vi_1 2.0623911143031277
rich_ii 2.051539523818834
coriolan 1.8181670293850258
much_ado 1.4567858550578254
all_well 1.4293029490903766
cymbelin 1.3719446855480473
t_night 1.0386120618788024
pericles 0
hen_viii 0


### Now on to the Motie files. 

We start with preparing the corpus, the index and the list of documents.

In [31]:
MOTIONS = "Moties/XML/MOT"

# prepare the files
prepare_data(MOTIONS, MOTIONS + "New", 0, 1)

# create the dinex
index_motions = index_collection(MOTIONS + "New")

index_with_freq_motions = add_frequencies(index_motions)

In [32]:
# Because there are so many motion files, we only print the first 100
doc_set_motions = make_doc_set(index_with_freq_motions)
print(list(doc_set_motions)[0:100])
print(len(doc_set_motions))

['2015035', '1808214', '2043301', '2046158', '1982352', '6198610', '1903281', '1833756', '1805622', '1986841', '5704433', '1878669', '1966228', '2031485', '1893239', '1912237', '1915951', '2019202', '6009308', '1826520', '1856046', '1859095', '2037886', '5769392', '1822452', '1852976', '2031737', '1920005', '1925072', '1981710', '1926528', '2021964', '2048088', '1851773', '1921758', '2040656', '1983409', '1822187', '1930975', '1859542', '1947819', '1983401', '2049938', '1967174', '1810268', '1928582', '1818050', '6179661', '1990920', '1963886', '1939806', '1828258', '6045820', '2059912', '6173184', '5753241', '1986151', '1905532', '1925692', '1810193', '6211020', '2049684', '1857392', '1890950', '1820964', '1890957', '1894274', '5645421', '1959636', '1978235', '6022527', '1953295', '1908067', '1857020', '1905283', '1927693', '5941862', '6289452', '1944543', '1929239', '1865418', '1854648', '2033252', '1824936', '1815925', '1927347', '1987408', '1819011', '5985506', '1951422', '1954417'

#### 8.1 Create a search engine for negation (very useful)

Since the collection is very large, we won't show the names of the documents, but we illustrate the correctness of our code with the numbers involved.

In [72]:
print("Documents not containing 'zonder': " + 
      str(len(negation_search(doc_set_motions, 
                              index_with_freq_motions, 
                              'zonder'))))

print("Documents containing 'zonder': " + 
      str(len(docs_with_token(index_with_freq_motions, 
                              'zonder'))))

print("Total number of documents in the corpus: " + 
      str(len(doc_set_motions)))

Documents not containing 'zonder': 26900
Documents containing 'newspaper': 1046
Total number of documents in the corpus: 27946


#### 8.2 Disjunction search

Input as a list of words. Output is unordered.

In [35]:
print(disjunction_search(['kort', 'macht'], index_with_freq_motions))

{'1894552', '2031240', '1907197', '1873652', '1969749', '2019052', '1890611', '1937152', '1951865', '1839057', '1854634', '5966588', '1932293', '1934445', '1872171', '1922178', '1841958', '1836208', '2000571', '1974612', '1845260', '1852819', '1910982', '1804890', '1834661', '1834384', '1873330', '1963559', '1810171', '1811263', '2061158', '2020711', '1839376', '1823866', '2042757', '1988356', '1959105', '1932226', '2043173', '1838354', '1814045', '1983455', '2058611', '1916534', '2060158', '1973726', '1858361', '1980304', '1831980', '1832750', '1950359', '1905096', '1841718', '1876259', '6289279', '1829990', '1965324', '1934815', '5578138', '1826894', '1835389', '1877709', '1835900', '1957349', '1832235', '2018939', '1958339', '1930868', '1952652', '1927026', '1834720', '1986127', '1957634', '1891808', '1812803', '1950410', '2031406', '6009234', '1907039', '1857699', '1991585', '1825393', '1978654', '2035249', '1838590', '1954504', '5578187', '1894882', '1871430', '1877752', '5753241'

##### 8.3.1 Ordering by number of terms matched

In [74]:
# We displayed this by length, because of the size
print(len(disj_search_simple_order(['naar', 'zoeken'], 
                               index_with_freq_motions)))
print('1851372:', num_of_terms(['naar', 'zoeken'], 
                               index_with_freq_motions, '1851372'))
print('5668949:', num_of_terms(['naar', 'zoeken'], 
                               index_with_freq_motions, '5668949'))

5259
1851372: 2
5668949: 1



##### 8.3.2 Ordering by term frequency

In [76]:
# Only print the first 30 because of the size of the corpus
count = 0

for doc in disj_search_term_freq_order(['uitzondering', 'zonder'], 
                                       index_with_freq_motions):
    if count < 30:
        print(doc, term_frequency(['uitzondering', 'zonder'], 
                               index_with_freq_motions, doc))
    count += 1

5752534 5
5858769 5
6165509 5
2036160 4
2062632 4
2045458 4
1990070 4
1917070 4
2027968 4
2063185 4
5701234 4
6224761 4
1898824 4
1871320 3
1921768 3
2037994 3
1898875 3
1921779 3
1818455 3
5692890 3
2019969 3
2056837 3
2029562 3
2015495 3
1898881 3
2056814 3
1992631 3
5863350 3
2015631 3
5869275 3


##### 8.3.3 Ordering by number of matched terms and term frequency

In [77]:
# Only print the first 30 due to the size of the corpus
count = 0

for doc in disj_search_combo_order(['uitzondering', 'zonder'], 
                                   index_with_freq_motions):
    if count < 30: 
        print (doc, tf_and_matchedNo_combo(['uitzondering', 'zonder'], 
                                       index_with_freq_motions, doc))
    count += 1

5752534 [2, 5]
1871320 [2, 3]
2056837 [2, 3]
2056814 [2, 3]
5863350 [2, 3]
6165521 [2, 2]
1811999 [2, 2]
1905968 [2, 2]
2007313 [2, 2]
2018000 [2, 2]
6177478 [2, 2]
2001958 [2, 2]
1920324 [2, 2]
1812045 [2, 2]
5858769 [1, 5]
6165509 [1, 5]
2036160 [1, 4]
2062632 [1, 4]
2045458 [1, 4]
1990070 [1, 4]
1917070 [1, 4]
2027968 [1, 4]
2063185 [1, 4]
5701234 [1, 4]
6224761 [1, 4]
1898824 [1, 4]
1921768 [1, 3]
2037994 [1, 3]
1898875 [1, 3]
1921779 [1, 3]


##### 8.4 Conjunctive search

Input is, again, a list of words.

In [34]:
print(conjunction_search(['kort','macht'], index_with_freq_motions))

{'1910310', '1910982', '1855695'}


#### 9.1 TF-IDF

In [79]:
returnval = TF_IDF('zonder', doc_set_motions, index_with_freq_motions)
for elem in returnval[:30]:
    print (elem)

[13.1412028440621, '6224761']
[13.1412028440621, '5752534']
[13.1412028440621, '5701234']
[13.1412028440621, '2063185']
[13.1412028440621, '2062632']
[13.1412028440621, '2027968']
[13.1412028440621, '1990070']
[13.1412028440621, '1917070']
[13.1412028440621, '1898824']
[9.855902133046575, '5869275']
[9.855902133046575, '5752624']
[9.855902133046575, '5692890']
[9.855902133046575, '2037994']
[9.855902133046575, '2029562']
[9.855902133046575, '2019969']
[9.855902133046575, '1992631']
[9.855902133046575, '1921779']
[9.855902133046575, '1921768']
[9.855902133046575, '1898881']
[9.855902133046575, '1898875']
[9.855902133046575, '1818455']
[6.57060142203105, '6313964']
[6.57060142203105, '6284030']
[6.57060142203105, '6199417']
[6.57060142203105, '6174948']
[6.57060142203105, '6172745']
[6.57060142203105, '6165504']
[6.57060142203105, '6165043']
[6.57060142203105, '6085273']
[6.57060142203105, '6085268']


#### 9.2 TF-IDF scoring

In [80]:
returnval = TF_IDF_search(['naar','gevaarlijk'], doc_set_motions, index_with_freq_motions)

for elem in returnval[:30]:
    print (elem)

('1876602', 28.58794760912831)
('1896402', 28.195211121541675)
('1826452', 21.440960706846234)
('2041356', 14.293973804564155)
('2062352', 13.508500829390881)
('2062405', 13.508500829390881)
('6081866', 12.212674713303658)
('2063081', 11.81993822571702)
('2062443', 11.81993822571702)
('1960847', 11.81993822571702)
('5778376', 11.81993822571702)
('1998806', 11.81993822571702)
('6081858', 10.524112109629797)
('1966248', 10.524112109629797)
('1997533', 10.13137562204316)
('2062475', 10.13137562204316)
('5731495', 10.13137562204316)
('1929440', 10.13137562204316)
('2062384', 10.13137562204316)
('1899883', 10.13137562204316)
('2042866', 10.13137562204316)
('1929425', 10.13137562204316)
('2005810', 10.13137562204316)
('1802743', 8.835549505955937)
('1917427', 8.4428130183693)
('1990639', 8.4428130183693)
('2061789', 8.4428130183693)
('1833138', 8.4428130183693)
('2043770', 8.4428130183693)
('1852543', 8.4428130183693)


#### 9.3 Euclidean length of docs

Function will return a dict, as that seems useful.

Careful! This thing will be very very long, which is why we have not included the result in the current status of this notebook. This is the beginning of the dict:

{'1822272': 19.131126469708992, '6324358': 16.822603841260722, '1969967': 26.13426869074396, '2062541': 26.795522013948524, '1927493': 22.627416997969522, '1863470': 20.97617696340303, '1883867': 27.147743920996454, '1884761': 22.847319317591726, '1924595': 25.592967784139454, '1807366': 32.07802986469088, '1978215': 19.72308292331602, '1969541': 35.482389998420345, '5941723': 29.274562336608895, '1922047': 21.047565179849187, '1872076': 21.02379604162864, '2052517': 32.78719262151, '1952563': 35.98610843089316, '1955718': 23.065125189341593, '5649055': 25.729360660537214, '1963911': 22.22611077089287, '5878188': 36.29049462324811, '1888724': 37.067505985701274, '2044403': 27.622454633866266, '1886567': 22.64950330581225, '2016526': 44.181444068749045, '1916534': 22.293496809607955, '2023929': 26.795522013948524, '1944418': 17.26267650163207, '1923937': 29.46183972531247, '1817433': 26.870057685088806, ......

In [82]:
print(euclidean_lengths(doc_set_motions, index_with_freq_motions))

#### 9.4 Cosine Similarity

*If a word appears multiple times in the query, the length of the query vector is not just the square root of the length of the query. However, as this would change all results by multiplying them with a constant, it would not interfere with the ranking.

In [83]:
result = cosine_similarity_search(['zonder', 'gevaar', 'naar'], 
                                  doc_set_motions, index_with_freq_motions)

# Show only 30 because of the size of the corpus

count = 0
for [doc,similarity] in result:
    if count < 30:
        print (doc, similarity)
    count += 1

6224761 0.4723507819680428
2062632 0.3448671363406695
2063185 0.3249946715053997
1929425 0.3194446745966354
1982261 0.31778464351192137
6196232 0.3142089859794301
1898824 0.3089696604389248
1929440 0.2966143295949211
2001180 0.2812579150222812
1917427 0.27782691651165714
1917070 0.2669106896335223
1898881 0.25838356481755875
1899429 0.25822499302351104
2027968 0.2576706440012177
2019969 0.2566739070868738
6235992 0.25394720885772337
6020986 0.251785993573431
1898875 0.24881978506648197
5847988 0.24636498010616054
1843575 0.24410815410035483
5752624 0.24263553770061774
2041946 0.24091613383605268
2020617 0.24062983644061398
1897105 0.24014038373472946
5701234 0.23922088344989634
1986176 0.2382984197241302
1921768 0.23813196787713978
1811954 0.23782436644357258
1909972 0.23562667933499212
5752534 0.2350393796307733


#### 9.5 Okapi BM25

We will need to do some calculations with the values of k1, b, the term frequency, the document vector length and the average document vector length. All of these are calculable except for k1 and b, which are free variables. The wikipedia page suggests putting b=0.75 and k1 being between 1.2 and 2, so we'll take b=0.75 and k1=1.6.

In [84]:
result = Okapi_BM25_search(['zonder', 'gevaar', 'naar'], 
                           doc_set_motions, index_with_freq_motions)
count = 0
for [doc, Okapi_score] in result:
    if count < 30:
        print (doc, Okapi_score)
    count += 1

1929425 3.123064218158644
1873014 3.07069714454952
1929440 3.065856887653787
1917427 3.039159563179651
2019969 3.032459236276599
1897105 2.988009944150866
1872513 2.9788001366348835
2033252 2.955692240669091
1872418 2.949558256855314
2055849 2.9398180834149983
1811954 2.9298859179264216
1876795 2.9214741236595105
1913813 2.898651581474115
5858761 2.8665191430646404
6064612 2.8477379989066534
5876031 2.8093005342931683
1879926 2.77516633826521
1899429 2.7714170903137116
1842779 2.7664597962116817
5649060 2.7344853842871792
1909972 2.7249852386700044
1852543 2.7095474300523543
1896948 2.703777897986984
1956993 2.695099298071142
1917070 2.690480196273434
6199417 2.6780444430061148
2022699 2.661200956359316
1988081 2.6540512390051014
6190285 2.6475819040399395
2063136 2.6448638690738506
